src/jit/gentree.cpp

```diff
@@ -9183,259 +9183,344 @@ GenTreePtr GenTree::GetChild(unsigned childNum)
-            {
-                return &m_node->AsStmt()->gtStmtExpr;
-            }
-            return nullptr;
+            case GT_FIELD:
+                m_advance = &GenTreeUseEdgeIterator::AdvanceField;
+                break;
 
-        case GT_ARR_ELEM:
-            if (m_state == 0)
-            {
-                return &m_node->AsArrElem()->gtArrObj;
-            }
-            else if (m_state <= m_node->AsArrElem()->gtArrRank)
-            {
-                return &m_node->AsArrElem()->gtArrInds[m_state - 1];
-            }
-            return nullptr;
+            case GT_STMT:
+                m_advance = &GenTreeUseEdgeIterator::AdvanceStmt;
+                break;
 
-        case GT_ARR_OFFSET:
-            switch (m_state)
-            {
-                case 0:
-                    return &m_node->AsArrOffs()->gtOffset;
-                case 1:
-                    return &m_node->AsArrOffs()->gtIndex;
-                case 2:
-                    return &m_node->AsArrOffs()->gtArrObj;
-                default:
-                    return nullptr;
-            }
+            case GT_ARR_ELEM:
+                m_advance = &GenTreeUseEdgeIterator::AdvanceArrElem;
+                break;
 
-        // Call, phi, and SIMD nodes are handled by MoveNext{Call,Phi,SIMD}UseEdge, repsectively.
-        case GT_CALL:
-        case GT_PHI:
-#ifdef FEATURE_SIMD
-        case GT_SIMD:
-#endif
-            break;
+            case GT_ARR_OFFSET:
+                m_advance = &GenTreeUseEdgeIterator::AdvanceArrOffset;
+                break;
+
+            case GT_DYN_BLK:
+                m_advance = &GenTreeUseEdgeIterator::AdvanceDynBlk;
+                break;
 
-        case GT_ASG:
+            case GT_STORE_DYN_BLK:
+                m_advance = &GenTreeUseEdgeIterator::AdvanceStoreDynBlk;
+                break;
+
+            case GT_CALL:
+                m_advance = &GenTreeUseEdgeIterator::AdvanceCall;
+                SetEntryStateForCall();
+                return;
+
+            default:
+                unreached();
+        }
+    }
+
+    // Advance to the first operand.
+    ++(*this);
+}
+
+void GenTreeUseEdgeIterator::AdvanceCmpXchg()
+{
+    FinishSimpleAdvance([this]() -> GenTree** {
+        switch (m_state)
+        {
+            case 0:
+                return &m_node->AsCmpXchg()->gtOpLocation;
+            case 1:
+                return &m_node->AsCmpXchg()->gtOpValue;
+            case 2:
+                return &m_node->AsCmpXchg()->gtOpComparand;
+            default:
+                return nullptr;
+        }
+    }());
+}
+
+void GenTreeUseEdgeIterator::AdvanceBoundsChk()
+{
+    FinishSimpleAdvance([this]() -> GenTree** {
+        switch (m_state)
+        {
+            case 0:
+                return &m_node->AsBoundsChk()->gtIndex;
+            case 1:
+                return &m_node->AsBoundsChk()->gtArrLen;
+            default:
+                return nullptr;
+        }
+    }());
+}
```

Comment:
> How would these all be implemented without lambda? It seems like the use of lambda in these cases is gratuitous.

The use of lambda allows for more concise code. The alternative is something like:
```
switch (m_state)
{
    case 0:
        m_edge = ...;
        break;

    case 1:
        m_edge = ...;
        break;

    case 2:
        m_edge = ...;
        break;

    default:
        m_state = -1;
        return;
}

if (*m_edge == nullptr)
{
    m_state = -1;
    return;
}

m_state++
```

Which is wordier (and was a more involved transformation from the original code). As it turns out, both MSVC and clang compile the versions present in this PR down to essentially what I've listed above: both the lambdas and `FinishSimpleAdvance` are inlined in all cases and the optimizer does a good job with the copies to `m_edge` and the modifications of `m_state`.

> In many cases you can just assert on the non-null-ness and simply do `m_state++`
If this turns out to be simple/easy (which it very well may with this redesign--it wasn't with the earlier design) then this seems like a good suggestion. I'll take a look.